experimental/split_mlir/src/iree/split_mlir/ExportFunctionsPassImpl.h

```diff
@@ -0,0 +1,106 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <fstream>
+#include <iterator>
+#include <regex>
+
+#include "iree/split_mlir/Passes.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/Support/raw_ostream.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/SymbolTable.h"
+
+namespace mlir {
+namespace iree {
+namespace split_mlir {
+
+#define GEN_PASS_DEF_EXPORTFUNCTIONS
+#include "iree/split_mlir/Passes.h.inc"  // IWYU pragma: export
+
+namespace {
+
+template <typename RegexIt>
+bool matchFunction(func::FuncOp op, std::cmatch& m,
+                   iterator_range<RegexIt> regexRange) {
+  StringRef name = op.getSymName();
+  for (const std::regex& regex : regexRange) {
+    if (std::regex_match(name.begin(), name.end(), m, regex)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+std::string& makeExportedFunctionFilePath(StringRef prefix,
+                                          StringRef functionName,
+                                          std::string& result) {
+  result.clear();
+  result += prefix;
+  result += functionName;
+  result += ".mlir";
+  return result;
+}
+
+void crearFunctionBody(func::FuncOp op) {
+  op.getFunctionBody().getBlocks().clear();
+  SymbolTable::setSymbolVisibility(op.getOperation(),
+                                   SymbolTable::Visibility::Nested);
+}
+
+struct ExportFunctionsPass
+    : public impl::ExportFunctionsBase<ExportFunctionsPass> {
+  using ExportFunctionsBase::ExportFunctionsBase;
+
+  LogicalResult initialize(MLIRContext* context) override {
+    std::transform(functions.begin(), functions.end(),
+                   std::back_inserter(regexList),
+                   [](const std::string& val) { return std::regex(val); });
+    return LogicalResult::success();
+  }
+
+  void runOnOperation() override {
+    mlir::ModuleOp moduleOp = getOperation();
+    SmallVector<func::FuncOp, 4> funcOps(
+        moduleOp.getBody()->getOps<func::FuncOp>().begin(),
+        moduleOp.getBody()->getOps<func::FuncOp>().end());
```

Comment:
The function body clearing doesn't look like it invalidates the iterator (I might be wrong, I haven't tried it). Is there a need to collect the functions in advance then?

Reply:
I changed it to walk the operation hierarchy. Not just to check the ops directly inside the module op.